provider/cluster/kube/cluster_ip_connections.go

```diff
@@ -0,0 +1,323 @@
+		_, err = c.ac.AkashV2beta1().ProviderLeasedIPs(c.ns).Update(ctx, &obj, metav1.UpdateOptions{})
+	} else {
+		_, err = c.ac.AkashV2beta1().ProviderLeasedIPs(c.ns).Create(ctx, &obj, metav1.CreateOptions{})
+	}
+
+	return err
+}
+
+func (c *client) PurgeDeclaredIPs(ctx context.Context, lID mtypes.LeaseID) error {
+	labelSelector := &strings.Builder{}
+	_, err := fmt.Fprintf(labelSelector, "%s=true,", builder.AkashManagedLabelName)
+	if err != nil {
+		return err
+	}
+	kubeSelectorForLease(labelSelector, lID)
+	result := c.ac.AkashV2beta1().ProviderLeasedIPs(c.ns).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+		LabelSelector: labelSelector.String(),
+	})
+
+	return result
+}
+
+func (c *client) ObserveIPState(ctx context.Context) (<-chan v1beta2.IPResourceEvent, error) {
+	var lastResourceVersion string
+	phpager := pager.New(func(ctx context.Context, opts metav1.ListOptions) (runtime.Object, error) {
+		resources, err := c.ac.AkashV2beta1().ProviderLeasedIPs(c.ns).List(ctx, opts)
+
+		if err == nil && len(resources.GetResourceVersion()) != 0 {
+			lastResourceVersion = resources.GetResourceVersion()
+		}
+		return resources, err
+	})
+
+	data := make([]akashtypes.ProviderLeasedIP, 0, 128)
+	err := phpager.EachListItem(ctx, metav1.ListOptions{}, func(obj runtime.Object) error {
+		plip := obj.(*akashtypes.ProviderLeasedIP)
+		data = append(data, *plip)
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	c.log.Info("starting ip passthrough watch", "resourceVersion", lastResourceVersion)
+	watcher, err := c.ac.AkashV2beta1().ProviderLeasedIPs(c.ns).Watch(ctx, metav1.ListOptions{
+		TypeMeta:             metav1.TypeMeta{},
+		LabelSelector:        "",
+		FieldSelector:        "",
+		Watch:                false,
+		AllowWatchBookmarks:  false,
+		ResourceVersion:      lastResourceVersion,
+		ResourceVersionMatch: "",
+		TimeoutSeconds:       nil,
+		Limit:                0,
+		Continue:             "",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	evData := make([]ipResourceEvent, len(data))
+	for i, v := range data {
+		ownerAddr, err := sdktypes.AccAddressFromBech32(v.Spec.LeaseID.Owner)
+		if err != nil {
+			return nil, err
+		}
+		providerAddr, err := sdktypes.AccAddressFromBech32(v.Spec.LeaseID.Provider)
+		if err != nil {
+			return nil, err
+		}
+
+		leaseID, err := v.Spec.LeaseID.ToAkash()
+		if err != nil {
+			return nil, err
+		}
+
+		proto, err := manifest.ParseServiceProtocol(v.Spec.Protocol)
+		if err != nil {
+			return nil, err
+		}
+
+		ev := ipResourceEvent{
+			eventType:    ctypes.ProviderResourceAdd,
+			lID:          leaseID,
+			serviceName:  v.Spec.ServiceName,
+			port:         v.Spec.Port,
+			externalPort: v.Spec.ExternalPort,
+			ownerAddr:    ownerAddr,
+			providerAddr: providerAddr,
+			sharingKey:   v.Spec.SharingKey,
+			protocol:     proto,
+		}
+		evData[i] = ev
+	}
+
+	data = nil
+
+	output := make(chan v1beta2.IPResourceEvent)
```

Comment:
any reason to not initialize it with a buffer size of `len(evData)` so that it doesn't initially block below?